File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    cnt = 0
    visited = [0]*n
    arr = [[] for _ in range(n)]

    for i in range(len(computers)):
        for j in range(len(computers[i])):
            if i==j:
                continue
            if computers[i][j] == 1:
                arr[i].append(j)


    def dfs(a):
        q=deque()
        q.append(a)

        while q:
            x = q.popleft()
            visited[x] = 1

            for i in arr[x]:
                if visited[i] == 0:
                    q.append(i)

    for i in range(len(visited)):
        if visited[i] == 1:
            continue
        dfs(i)
        cnt += 1

    return cnt

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static int[] visited;
    static List<List<Integer>> arr;
    public static void bfs(int k){
        visited[k]=1;

        for(int i : arr.get(k)){
            if(visited[i]==0){
                bfs(i);
            }
        }
    }
    public static int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new int[n];
        arr = new ArrayList<>();

        for(int i=0; i<n; i++){
            List<Integer> row = new ArrayList<>();
            for(int j=0; j<n; j++){
                if(i!=j && computers[i][j]==1){
                    row.add(j);
                }
            }
            arr.add(row);
        }

        for(int k=0; k<n; k++){
            if(visited[k]==0){
                bfs(k);
                answer += 1;
            }
        }

        return answer;
    }
}